File: js/ai/Pusher.js
//Translated from pusher.rb using babyruby2js
'use strict';

var main = require('../main');
var Grid = require('../Grid');
var Heuristic = require('./Heuristic');
var inherits = require('util').inherits;

var sOK = main.sOK;


/** @class
 *  Way of "pushing" our influence further...
 *  Still very naive; for that reason the coeff are rather low.
 */
function Pusher(player) {
    Heuristic.call(this, player);
    this.allyCoeff = this.getGene('ally-infl', 0.03, 0.01, 1.0);
    this.enemyCoeff = this.getGene('enemy-infl', 0.13, 0.01, 1.0);
}
inherits(Pusher, Heuristic);
module.exports = Pusher;

Pusher.prototype.evalBoard = function (stateYx, scoreYx) {
    var myScoreYx = this.scoreGrid.yx;
    for (var j = 1; j <= this.gsize; j++) {
        for (var i = 1; i <= this.gsize; i++) {
            if (stateYx[j][i] < sOK) continue;
            var score = myScoreYx[j][i] = this.evalMove(i, j);
            scoreYx[j][i] += score;
        }
    }
};

Pusher.prototype.evalMove = function (i, j) {
    var inf = this.inf.map[j][i];
    var enemyInf = inf[this.enemyColor];
    var allyInf = inf[this.color];
    if (enemyInf === 0 || allyInf === 0) {
        return 0;
    }
    if (!this.canConnect(i, j, this.color)) return 0;

    var fillTer = this.enemyTerritoryScore(i, j, this.color);
    if (fillTer < 0) fillTer = 0; // Spacer will count <0 scores

    var score = fillTer + this.enemyCoeff * enemyInf - this.allyCoeff * allyInf;
    if (main.debug) {
        main.log.debug('Pusher heuristic sees influences ' + allyInf + ' - ' + enemyInf + ' at ' + i + ',' + j + ' -> ' + '%.03f'.format(score));
    }
    return score;
};

File: rb/ai/pusher.rb
# Quite a dumb way of "pushing" our influence further...
# For that reason the coeff are rather low.
# This should eventually disappear.

require_relative "heuristic"

class Pusher < Heuristic

  def initialize(player)
    super
    @ally_coeff = get_gene("ally-infl", 0.1, 0.01, 4.0)
    @enemy_coeff = get_gene("enemy-infl", 0.4, 0.01, 4.0)
  end

  def eval_move(i,j)
    inf = @inf.map[j][i]
    enemy_inf = inf[@enemy_color]
    ally_inf = inf[@color]
    
    return 0 if enemy_inf == 0 or ally_inf == 0
    score = 0.33 * (@enemy_coeff * enemy_inf - @ally_coeff * ally_inf)
    $log.debug("Pusher heuristic sees influences #{ally_inf} - #{enemy_inf} at #{i},#{j} -> #{'%.03f' % score}") if $debug
    return score
  end

end
